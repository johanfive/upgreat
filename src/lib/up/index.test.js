const fs = require('../../util/fs')
jest.mock('../../util/fs')
const json = require('../../util/json')
jest.mock('../../util/json')
const runTests = require('./runTests')
jest.mock('./runTests')
const upgrade = require('./upgrade')
jest.mock('./upgrade')

const Up = require('./')

describe('run', () => {
  it('should fail if plan cannot be read', async () => {
    json.readJSONFile.mockImplementationOnce(() =>
      Promise.reject(new Error('the message')),
    )

    const up = new Up([])

    await expect(up.run()).rejects.toMatchSnapshot()
    expect(json.readJSONFile.mock.calls).toMatchSnapshot()
  })

  it('should fail if initial tests dont pass', async () => {
    json.readJSONFile.mockImplementationOnce(() => Promise.resolve())
    runTests.mockImplementationOnce(() =>
      Promise.reject({
        stderr: 'stderr text',
      }),
    )

    const up = new Up([])
    const spy = jest.spyOn(up, 'log')
    spy.mockImplementation(() => {})

    await expect(up.run()).rejects.toMatchSnapshot()
    expect(spy.mock.calls).toMatchSnapshot()
  })

  it('should do upgrades in order', async () => {
    json.readJSONFile.mockImplementationOnce(() =>
      Promise.resolve([{ name: 'p1' }, { name: 'p2' }, { name: 'p3' }]),
    )
    runTests.mockImplementationOnce(() => Promise.resolve())
    const upgradeFn = jest
      .fn()
      .mockImplementationOnce(() =>
        Promise.resolve({
          name: 'p1',
        }),
      )
      .mockImplementationOnce(() =>
        Promise.resolve({
          name: 'p2',
          err: 'some err',
        }),
      )
      .mockImplementationOnce(() =>
        Promise.resolve({
          name: 'p3',
        }),
      )
    upgrade.mockImplementationOnce(() => upgradeFn)

    const up = new Up([])
    const spy = jest.spyOn(up, 'log')
    spy.mockImplementation(() => {})

    await up.run()

    expect(spy.mock.calls).toMatchSnapshot()
  })
})
