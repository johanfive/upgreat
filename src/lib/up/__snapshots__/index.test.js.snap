// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run should do upgrades in order 1`] = `
Array [
  Array [
    "got plan from .upgreat/plan.json",
  ],
  Array [
    "running tests before starting upgrades...",
  ],
  Array [
    "👍  tests are fine, starting upgrades",
  ],
  Array [
    "✅  upgraded 2",
  ],
  Array [
    "❌  errors 1",
  ],
]
`;

exports[`run should fail if initial tests dont pass 1`] = `[Error: your tests need to pass before attempting the upgrade! 😒]`;

exports[`run should fail if initial tests dont pass 2`] = `
Array [
  Array [
    "got plan from .upgreat/plan.json",
  ],
  Array [
    "running tests before starting upgrades...",
  ],
]
`;

exports[`run should fail if plan cannot be read 1`] = `[Error: the message]`;

exports[`run should fail if plan cannot be read 2`] = `
Array [
  Array [
    ".upgreat/plan.json",
  ],
]
`;
